File: claude_long/python/progress_manager.py
"""
Progress Manager - 进度管理器 (Python 实现)

管理 claude-progress.txt 文件，记录和读取项目会话进度。
"""

import os
from datetime import datetime
from typing import List, Dict, Optional


class ProgressManager:
    """管理项目进度文件的类"""

    def __init__(self, progress_file: str = "claude-progress.txt"):
        self.progress_file = progress_file

    def initialize(self) -> None:
        """初始化进度文件"""
        if os.path.exists(self.progress_file):
            print(f"Warning: {self.progress_file} already exists")
            return

        initial_content = """# Claude Long-Running Agent Progress Log

This file tracks the progress of long-running agent sessions.

=== Session 1: Initial Setup ===
Completed:
- Initialized progress tracking system

Status: Ready to begin development

---
"""
        with open(self.progress_file, 'w', encoding='utf-8') as f:
            f.write(initial_content)

        print(f"Initialized progress file: {self.progress_file}")

    def read(self) -> str:
        """读取整个进度文件"""
        if not os.path.exists(self.progress_file):
            return ""
        with open(self.progress_file, 'r', encoding='utf-8') as f:
            return f.read()

    def get_recent_sessions(self, count: int = 5) -> List[str]:
        """获取最近的几个会话记录"""
        content = self.read()
        sessions = content.split('=== Session')
        if len(sessions) > 0 and 'Progress Log' in sessions[0]:
            sessions = sessions[1:]
        recent = sessions[-count:] if len(sessions) > count else sessions
        return ['=== Session' + s for s in recent]

    def add_session(
        self,
        session_number: Optional[int] = None,
        completed_work: List[str] = None,
        status: str = "",
        next_steps: List[str] = None,
        notes: str = ""
    ) -> None:
        """添加新的会话记录"""
        if completed_work is None:
            completed_work = []
        if next_steps is None:
            next_steps = []

        if session_number is None:
            content = self.read()
            session_number = content.count('=== Session') + 1

        timestamp = datetime.now().strftime("%Y-%m-%d %H:%M")
        session_entry = f"\n=== Session {session_number}: {timestamp} ===\n"

        if completed_work:
            session_entry += "\nCompleted:\n"
            for work in completed_work:
                session_entry += f"- {work}\n"

        if status:
            session_entry += f"\nStatus: {status}\n"

        if next_steps:
            session_entry += "\nNext Steps:\n"
            for step in next_steps:
                session_entry += f"- {step}\n"

        if notes:
            session_entry += f"\nNotes:\n{notes}\n"

        session_entry += "\n---\n"

        with open(self.progress_file, 'a', encoding='utf-8') as f:
            f.write(session_entry)

        print(f"Added Session {session_number} to {self.progress_file}")


if __name__ == "__main__":
    import sys
    if len(sys.argv) < 2:
        print("Usage: python progress_manager.py <command>")
        print("Commands: init, read, recent, add")
        sys.exit(1)

    pm = ProgressManager()
    command = sys.argv[1]

    if command == "init":
        pm.initialize()
    elif command == "read":
        print(pm.read())
    elif command == "recent":
        count = int(sys.argv[2]) if len(sys.argv) > 2 else 5
        for session in pm.get_recent_sessions(count):
            print(session)
    elif command == "add":
        message = ' '.join(sys.argv[2:]) if len(sys.argv) > 2 else "Session completed"
        pm.add_session(completed_work=[message], status="Session completed")

File: claude_long/go/progress_manager.go
package main

import (
	"fmt"
	"os"
	"strings"
	"time"
)

// ProgressManager 管理项目进度文件
type ProgressManager struct {
	ProgressFile string
}

// Session 表示一个会话记录
type Session struct {
	SessionNumber int
	Date          time.Time
	CompletedWork []string
	Status        string
	NextSteps     []string
	Notes         string
}

// NewProgressManager 创建新的进度管理器
func NewProgressManager(progressFile string) *ProgressManager {
	if progressFile == "" {
		progressFile = "claude-progress.txt"
	}
	return &ProgressManager{
		ProgressFile: progressFile,
	}
}

// Initialize 初始化进度文件
func (pm *ProgressManager) Initialize() error {
	if _, err := os.Stat(pm.ProgressFile); err == nil {
		fmt.Printf("Warning: %s already exists\n", pm.ProgressFile)
		return nil
	}

	initialContent := `# Claude Long-Running Agent Progress Log

This file tracks the progress of long-running agent sessions.
Each session should record:
- What was accomplished
- Current status
- Any issues or blockers
- Next steps

格式：
=== Session N: YYYY-MM-DD HH:MM ===
Completed:
- Task 1
- Task 2

Status: Current project state
Next: What needs to be done next

---
`

	err := os.WriteFile(pm.ProgressFile, []byte(initialContent), 0644)
	if err != nil {
		return fmt.Errorf("failed to initialize progress file: %w", err)
	}

	fmt.Printf("Initialized progress file: %s\n", pm.ProgressFile)
	return nil
}

// Read 读取整个进度文件
func (pm *ProgressManager) Read() (string, error) {
	data, err := os.ReadFile(pm.ProgressFile)
	if err != nil {
		if os.IsNotExist(err) {
			return "", nil
		}
		return "", err
	}
	return string(data), nil
}

// GetRecentSessions 获取最近的几个会话记录
func (pm *ProgressManager) GetRecentSessions(count int) ([]string, error) {
	content, err := pm.Read()
	if err != nil {
		return nil, err
	}

	sessions := strings.Split(content, "=== Session")

	// 移除文件头部说明
	if len(sessions) > 0 && strings.Contains(sessions[0], "Progress Log") {
		sessions = sessions[1:]
	}

	// 获取最近的 N 个会话
	start := len(sessions) - count
	if start < 0 {
		start = 0
	}
	recent := sessions[start:]

	// 重新添加会话标记
	result := make([]string, len(recent))
	for i, s := range recent {
		result[i] = "=== Session" + s
	}

	return result, nil
}

// AddSession 添加新的会话记录
func (pm *ProgressManager) AddSession(session Session) error {
	// 自动确定会话编号
	if session.SessionNumber == 0 {
		content, _ := pm.Read()
		session.SessionNumber = strings.Count(content, "=== Session") + 1
	}

	// 构建会话记录
	var sb strings.Builder
	sb.WriteString(fmt.Sprintf("\n=== Session %d: %s ===\n",
		session.SessionNumber,
		session.Date.Format("2006-01-02 15:04")))

	if len(session.CompletedWork) > 0 {
		sb.WriteString("\nCompleted:\n")
		for _, work := range session.CompletedWork {
			sb.WriteString(fmt.Sprintf("- %s\n", work))
		}
	}

	if session.Status != "" {
		sb.WriteString(fmt.Sprintf("\nStatus: %s\n", session.Status))
	}

	if len(session.NextSteps) > 0 {
		sb.WriteString("\nNext Steps:\n")
		for _, step := range session.NextSteps {
			sb.WriteString(fmt.Sprintf("- %s\n", step))
		}
	}

	if session.Notes != "" {
		sb.WriteString(fmt.Sprintf("\nNotes:\n%s\n", session.Notes))
	}

	sb.WriteString("\n---\n")

	// 追加到文件
	f, err := os.OpenFile(pm.ProgressFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return err
	}
	defer f.Close()

	_, err = f.WriteString(sb.String())
	if err != nil {
		return err
	}

	fmt.Printf("Added Session %d to %s\n", session.SessionNumber, pm.ProgressFile)
	return nil
}

// GetSummary 获取进度摘要
func (pm *ProgressManager) GetSummary() (map[string]interface{}, error) {
	content, err := pm.Read()
	if err != nil {
		return nil, err
	}

	sessionCount := strings.Count(content, "=== Session")
	recentSessions, _ := pm.GetRecentSessions(3)

	fileInfo, _ := os.Stat(pm.ProgressFile)
	var fileSize int64 = 0
	if fileInfo != nil {
		fileSize = fileInfo.Size()
	}

	return map[string]interface{}{
		"total_sessions":  sessionCount,
		"recent_sessions": recentSessions,
		"file_size":       fileSize,
	}, nil
}

// ArchiveOldSessions 归档旧的会话记录
func (pm *ProgressManager) ArchiveOldSessions(keepRecent int) error {
	content, err := pm.Read()
	if err != nil {
		return err
	}

	sessions := strings.Split(content, "=== Session")

	// 保留文件头部说明
	header := ""
	if len(sessions) > 0 && strings.Contains(sessions[0], "Progress Log") {
		header = sessions[0]
		sessions = sessions[1:]
	}

	if len(sessions) <= keepRecent {
		fmt.Println("No sessions to archive")
		return nil
	}

	// 分离要归档的和要保留的
	toArchive := sessions[:len(sessions)-keepRecent]
	toKeep := sessions[len(sessions)-keepRecent:]

	// 创建归档文件
	archiveFile := strings.Replace(pm.ProgressFile, ".txt", "-archive.txt", 1)
	timestamp := time.Now().Format("20060102-150405")

	archiveContent := fmt.Sprintf("\n\n# Archived on %s\n\n", timestamp)
	archiveContent += "=== Session" + strings.Join(toArchive, "=== Session")

	// 追加到归档文件
	f, err := os.OpenFile(archiveFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return err
	}
	defer f.Close()
	f.WriteString(archiveContent)

	// 重写主文件
	newContent := header + "=== Session" + strings.Join(toKeep, "=== Session")
	err = os.WriteFile(pm.ProgressFile, []byte(newContent), 0644)
	if err != nil {
		return err
	}

	fmt.Printf("Archived %d sessions to %s\n", len(toArchive), archiveFile)
	fmt.Printf("Kept %d recent sessions in %s\n", len(toKeep), pm.ProgressFile)

	return nil
}
